File: 0020.有效的括号/0020-有效的括号.py
class Solution:
    def isValid(self, s: str) -> bool:
        if len(s)%2!=0:
            return False
        stack=[]
        for i in s:
            if i=="(" or i=="[" or i=="{":
                stack.append(i)
            else:
                if len(stack)==0:
                    return False
                else:
                    temp=stack.pop()
                    if i==")":
                        if temp!='(':
                            return False
                    elif i=="]":
                        if temp!='[':
                            return False   
                    elif i=="}":
                        if temp!='{':
                            return False     
        if len(stack)==0:
            return True
        else:
            return False                

File: 0020.有效的括号/0020-有效的括号.cpp
class Solution {
public:
    bool isValid(string s) {
        stack<int> st;
        if( s.size()%2 != 0 ) return false;
        for( int i = 0; i < s.size(); i++){
            if( s[i] == '(') st.push(')');
            else if( s[i] == '{') st.push('}');
            else if( s[i] == '[') st.push(']');\
            else if( st.empty() || s[i] != st.top()) return false;
            else if( st.top() == s[i]) st.pop();
        }
        return st.empty();

    }
};